rllib/agents/ppo/ppo_tf_policy.py

```diff
@@ -385,61 +53,204 @@ def setup_config(
+            # if necessary.
+            if self.config["kl_coeff"] > 0.0:
+                total_loss += self.kl_coeff * mean_kl_loss
+
+            # Store stats in policy for stats_fn.
+            self._total_loss = total_loss
+            self._mean_policy_loss = mean_policy_loss
+            self._mean_vf_loss = mean_vf_loss
+            self._mean_entropy = mean_entropy
+            # Backward compatibility: Deprecate self._mean_kl.
+            self._mean_kl_loss = self._mean_kl = mean_kl_loss
+            self._value_fn_out = value_fn_out
+
+            return total_loss
+
+        @override(base)
+        def postprocess_trajectory(
+            self, sample_batch, other_agent_batches=None, episode=None
+        ):
+            sample_batch = super().postprocess_trajectory(sample_batch)
+            return compute_gae_for_sample_batch(
+                self, sample_batch, other_agent_batches, episode
+            )
+
+        @override(base)
+        def compute_gradients_fn(
+            self, optimizer: LocalOptimizer, loss: TensorType
+        ) -> ModelGradients:
+            # Compute the gradients.
+            variables = self.model.trainable_variables
+            if isinstance(self.model, ModelV2):
+                variables = variables()
+            grads_and_vars = optimizer.compute_gradients(loss, variables)
+
+            # Clip by global norm, if necessary.
+            if self.config["grad_clip"] is not None:
+                # Defuse inf gradients (due to super large losses).
+                grads = [g for (g, v) in grads_and_vars]
+                grads, _ = tf.clip_by_global_norm(grads, self.config["grad_clip"])
+                # If the global_norm is inf -> All grads will be NaN. Stabilize this
+                # here by setting them to 0.0. This will simply ignore destructive loss
+                # calculations.
+                self.grads = [tf.where(tf.math.is_nan(g), tf.zeros_like(g), g) for g in grads]
+                clipped_grads_and_vars = list(zip(self.grads, variables))
+                return clipped_grads_and_vars
+            else:
+                return grads_and_vars
+
+        @override(base)
+        def stats_fn(self, train_batch: SampleBatch) -> Dict[str, TensorType]:
+            return {
+                "cur_kl_coeff": tf.cast(self.kl_coeff, tf.float64),
+                "cur_lr": tf.cast(self.cur_lr, tf.float64),
+                "total_loss": self._total_loss,
+                "policy_loss": self._mean_policy_loss,
+                "vf_loss": self._mean_vf_loss,
+                "vf_explained_var": explained_variance(
+                    train_batch[Postprocessing.VALUE_TARGETS], self._value_fn_out
+                ),
+                "kl": self._mean_kl_loss,
+                "entropy": self._mean_entropy,
+                "entropy_coeff": tf.cast(self.entropy_coeff, tf.float64),
+            }
+
+    return PPOTFPolicy
+
+
+PPODynamicTFPolicy = get_ppo_tf_policy(DynamicTFPolicyV2)
```

Comment:
Should we just name `DynamicTFPolicyV2` into `TFPolicyV2`?

I feel like there is no point in making TF policies not "dynamic" (which basically only means automatic loss graph creation). Let me know what you think.

Reply:
I guess so. probably nobody is using TFPolicy directly.
we should definitely deprecate and do this with a separate PR though.